File: 0042.Trapping Rain Water/20_11_11_two_pointers_ac.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();

        if(n == 0) return 0;

        int res = 0;
        int left = 0, right = n - 1;
        int left_max = height[0], right_max = height[n-1];

        while(left <= right) {
            left_max = max(left_max, height[left]);
            right_max = max(right_max, height[right]);

            if(left_max < right_max) {
                res += (left_max - height[left]);
                left ++;
            } else {
                res += (right_max - height[right]);
                right --;
            }
        }

        return res;
    }
};

int main() {
    Solution sol;
    vector<int> v = {0,1,0,2,1,0,1,3,2,1,2,1};
    std::cout << sol.trap(v);
    return 0;
}

File: 0042.Trapping Rain Water/20_11_11_two_pointers_ac.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)

        # 边界条件
        if n == 0:
            return 0

        res = 0

        # 初始化左右指针
        left, right = 0, n - 1
        # max_left 记录 height[0..left] 中最高柱子高度
        # max_right 记录 height[right..end] 中最高柱子高度
        max_left, max_right = height[left], height[right]

        while left < right:
            # 更新 max_left, max_right
            if height[left] > max_left:
                max_left = height[left]
            if height[right] > max_right:
                max_right = height[right]

            # 基于短板效应，选取最高柱子高度较低一侧的可达水柱高度加入结果
            if max_left < max_right:
                res += max_left - height[left]
                left += 1
            else:
                res += max_right - height[right]
                right -= 1

        return res